runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/Environments.java

```diff
@@ -117,8 +121,12 @@ public static JavaVersion forSpecification(String specification) {
           return ver;
         }
       }
-      throw new UnsupportedOperationException(
-          String.format("unsupported Java version: %s", specification));
+      JavaVersion fallback = java11;
```

Comment:
Could we have two methods here, where one is "getJavaVersionOrDefault" or similar? There are some cases where we probably want to have a strict check (i.e. expecting translatation from one of the existing enums, looking for a specific supported Docker container version where defaulting can lead to confusing behavior)

Reply:
Sure, I will add that.
Do you happen to already know some places where we would use the stricter version ?

Reply:
I added a strict method `forSpecificationStrict`.

the java version ends up being used in
* `JAVA_SDK_HARNESS_CONTAINER_URL` for which I wan to use the lenient version for now
* `DataflowRunner.fromOptions` for which I wan to use the lenient version
* `DataflowRunner.getDefaultContainerImageNameForJob` for which I wan to use the lenient version